File: Python/programmers/etc/level2/뒤에_있는_큰_수_찾기.py
def solution(numbers):
    dp = [-1]*len(numbers)
    stack = []
    for idx, number in enumerate(numbers):
        while stack and numbers[stack[-1]] < number:
            dp[stack.pop()] = number
        stack.append(idx)
    return dp


numbers = [2, 3, 3, 5]
print(solution(numbers)) # [3, 5, 5, -1]

numbers = [9, 1, 5, 3, 6, 2]
print(solution(numbers)) # [-1, 5, 6, 6, -1, -1]

File: Java/programmers/etc/level2/뒤에있는큰수찾기.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 뒤에있는큰수찾기 {
    public int[] solution(int[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        Arrays.fill(answer, -1);

        Stack<Integer> stk = new Stack<>();
        for (int i=0; i<n; i++) {
            while (!stk.isEmpty() && numbers[stk.peek()] < numbers[i]) {
                answer[stk.pop()] = numbers[i];
            }
            stk.add(i);
        }
        
        return answer;
    }
}
